framework/src/convergence/ReferenceResidualConvergence.C

```diff
@@ -0,0 +1,524 @@
+          for (unsigned int var_num = 0; var_num < s.n_vars(); var_num++)
+            if (_group_variables[i][j] == s.variable_name(var_num))
+            {
+              if (nonlinear_sys.isScalarVariable(_soln_vars[var_num]))
+                ++num_scalar_vars;
+              else
+                ++num_field_vars;
+              break;
+            }
+      }
+      if (num_scalar_vars > 0 && num_field_vars > 0)
+        mooseWarning("In the 'group_variables' parameter, standard variables and scalar variables "
+                     "are grouped together in group ",
+                     i);
+    }
+  }
+
+  for (unsigned int i = 0; i < n_soln_vars; ++i)
+  {
+    if (_group_soln_var_names[_variable_group_num_index[i]].empty())
+    {
+      _group_soln_var_names[_variable_group_num_index[i]] = _soln_var_names[i];
+      if (_use_group_variables && _variable_group_num_index[i] < _group_variables.size())
+        _group_soln_var_names[_variable_group_num_index[i]] += " (grouped) ";
+    }
+
+    if (!_reference_vector && _group_ref_resid_var_names[_variable_group_num_index[i]].empty())
+    {
+      _group_ref_resid_var_names[_variable_group_num_index[i]] = _ref_resid_var_names[i];
+      if (_use_group_variables && _variable_group_num_index[i] < _group_variables.size())
+        _group_ref_resid_var_names[_variable_group_num_index[i]] += " (grouped) ";
+    }
+  }
+
+  if (!_reference_vector)
+  {
+    const unsigned int size_soln_vars = _soln_vars.size();
+    _scaling_factors.resize(size_soln_vars);
+    for (unsigned int i = 0; i < size_soln_vars; ++i)
+      if (nonlinear_sys.isScalarVariable(_soln_vars[i]))
+        _scaling_factors[i] = nonlinear_sys.getScalarVariable(0, _soln_vars[i]).scalingFactor();
+      else
+        _scaling_factors[i] = nonlinear_sys.getVariable(/*tid*/ 0, _soln_vars[i]).scalingFactor();
+  }
+  _initialized = true;
+}
+
+void
+ReferenceResidualConvergence::updateReferenceResidual()
+{
+  NonlinearSystemBase & _current_nl_sys = _fe_problem.currentNonlinearSystem();
+  AuxiliarySystem & aux_sys = _fe_problem.getAuxiliarySystem();
+  System & s = _current_nl_sys.system();
+  auto & as = aux_sys.sys();
+
+  std::fill(_group_resid.begin(), _group_resid.end(), 0.0);
+  std::fill(_group_output_resid.begin(), _group_output_resid.end(), 0.0);
+  if (_local_norm)
+    std::fill(_group_ref_resid.begin(), _group_ref_resid.end(), 1.0);
+  else
+    std::fill(_group_ref_resid.begin(), _group_ref_resid.end(), 0.0);
+
+  for (const auto i : index_range(_soln_vars))
+  {
+    Real resid = 0.0;
+    const auto group = _variable_group_num_index[i];
+    if (_local_norm)
+    {
+      mooseAssert(_current_nl_sys.RHS().size() == (*_reference_vector).size(),
+                  "Sizes of nonlinear RHS and reference vector should be the same.");
+      mooseAssert((*_reference_vector).size(), "Reference vector must be provided.");
+      // Add a tiny number to the reference to prevent a divide by zero.
```

Comment:
I don't think so. I tried following the pointwise divide in PETSc but couldn't really find the actual divide in the end, so I don't know if it has some kind of assert to prevent division by zero.

I don't know their reasoning and this line might not be bulletproof, but it does guard against the much more likely case that the vector is zero than if the vector has a negative entry of size `-std::numeric_limits<Number>::min()`.